File: 42.trapping-rain-water.py
#
# @lc app=leetcode id=42 lang=python3
#
# [42] Trapping Rain Water
#


# @lc code=start
class Solution:
    def trap(self, height: List[int]) -> int:
        if len(height) <= 2:
            return 0
        # two pointers
        l, r = 0, 0
        n = len(height)
        sections = []
        # finding sections that trap water
        while r < n:
            if l == r:
                r += 1
                continue
            if height[r] < height[l]:
                r += 1
            else:
                interval = height[l : r + 1]
                if len(interval) > 2:
                    sections.append(interval)
                l = r
        # for the rest part that did not parse, do it reversely to find rest sections
        if l != n:
            rest = list(reversed(height[l:]))
            l, r = 0, 0
            n = len(rest)
            while r < n:
                if l == r:
                    r += 1
                    continue
                if rest[r] < rest[l]:
                    r += 1
                else:
                    interval = rest[l : r + 1]
                    if len(interval) > 2:
                        sections.append(interval)
                    l = r
        # calculate answer
        ans = 0
        for s in sections:
            waterLine = min(s[0], s[-1])
            for val in s:
                ans += max(0, waterLine - val)
        return ans


# @lc code=end

File: 42.trapping-rain-water.go
/*
 * @lc app=leetcode id=42 lang=golang
 *
 * [42] Trapping Rain Water
 */

// @lc code=start
package main

import (
	"slices"
)

/*
Finds trapping intervals
*/
func trap(height []int) int {
	if len(height) <= 2 {
		return 0
	}
	var count func(height []int) []int
	count = func(height []int) []int {
		// 2 pointers
		intervals := [][]int{}
		n := len(height)
		l, r := 0, 0
		for r < n {
			if l == r {
				r++
				continue
			}
			if height[r] < height[l] {
				r++
			} else {
				interval := []int{l, r}
				if r > l+1 {
					intervals = append(intervals, interval)
				}
				l = r
			}
		}
		// count trap volume
		ans := 0
		for _, interval := range intervals {
			left := height[interval[0]]
			right := height[interval[1]]
			waterline := min(left, right)
			for i := interval[0] + 1; i < interval[1]; i++ {
				ans += waterline - height[i]
			}
		}
		return []int{ans, l}
	}
	ans1 := count(height)
	// if l not reach end, reverse the rest and keep finding intervals
	if ans1[1] != len(height)-1 {
		res := height[ans1[1]:]
		slices.Reverse(res)
		return ans1[0] + count(res)[0]
	}
	return ans1[0]
}

func min(x, y int) int {
	if x < y {
		return x
	}
	return y
}

// @lc code=end
